File: benchmarks/bench_masking.cpp
// bench_masking.cpp — RFC 6455 mask/unmask throughput

#include "bench_common.hpp"
#include <wscpp/frame/parser.hpp>
#include <wscpp/frame/builder.hpp>
#include <array>
#include <cstdio>
#include <vector>

using namespace wscpp;
using namespace wscpp::bench;

namespace {

void mask_payload(std::vector<uint8_t>& data, const std::array<uint8_t, 4>& key) {
    for (std::size_t i = 0; i < data.size(); ++i) {
        data[i] ^= key[i % 4];
    }
}

} // namespace

int main() {
    const std::size_t payload_size = 1024 * 1024;
    const int iterations = 200;

    std::vector<uint8_t> payload(payload_size);
    for (std::size_t i = 0; i < payload_size; ++i) {
        payload[i] = static_cast<uint8_t>(i & 0xFF);
    }

    const std::array<uint8_t, 4> key = {{0x37, 0xfa, 0x21, 0x3d}};

    const clock::time_point mask_start = clock::now();
    for (int i = 0; i < iterations; ++i) {
        mask_payload(payload, key);
        mask_payload(payload, key);
    }
    const clock::time_point mask_end = clock::now();

    const double mask_sec = elapsed_sec(mask_start, mask_end);
    print_throughput("mask_unmask_xor", static_cast<double>(payload_size * iterations * 2),
                     mask_sec, iterations * 2);

    frame::builder builder;
    frame::parser parser;
    frame::frame_header header;
    std::vector<uint8_t> parsed;

    const clock::time_point roundtrip_start = clock::now();
    for (int i = 0; i < 50; ++i) {
        const std::vector<uint8_t> frame =
            builder.build_binary(payload.data(), payload.size(), true, true);
        parser.reset();
        if (parser.parse(frame.data(), frame.size(), header, parsed) !=
            frame::parse_result::COMPLETE) {
            std::fprintf(stderr, "masked parse failed\n");
            return 1;
        }
    }
    const clock::time_point roundtrip_end = clock::now();

    const double rt_sec = elapsed_sec(roundtrip_start, roundtrip_end);
    print_throughput("masked_build_parse", static_cast<double>(payload_size * 50),
                     rt_sec, 50);

    return 0;
}

File: rust/ws-rs-benches/src/bin/bench_masking.rs
// bench_masking — RFC 6455 mask/unmask throughput (mirrors C++ bench_masking)

use std::time::Instant;

use ws_rs::frame::{apply_mask, FrameBuilder, ParseResult, Parser};
use ws_rs_benches::common::print_throughput;

fn main() {
    const PAYLOAD_SIZE: usize = 1024 * 1024;
    const MASK_ITERATIONS: i32 = 200;
    const ROUNDTRIP_ITERATIONS: i32 = 50;
    const KEY: [u8; 4] = [0x37, 0xfa, 0x21, 0x3d];

    let mut payload = vec![0u8; PAYLOAD_SIZE];
    for (i, b) in payload.iter_mut().enumerate() {
        *b = (i & 0xFF) as u8;
    }

    let mask_start = Instant::now();
    for _ in 0..MASK_ITERATIONS {
        apply_mask(&mut payload, &KEY);
        apply_mask(&mut payload, &KEY);
    }
    let mask_sec = ws_rs_benches::common::elapsed_sec(mask_start, Instant::now());
    print_throughput(
        "mask_unmask_xor",
        PAYLOAD_SIZE as f64 * MASK_ITERATIONS as f64 * 2.0,
        mask_sec,
        MASK_ITERATIONS * 2,
    );

    let roundtrip_start = Instant::now();
    for _ in 0..ROUNDTRIP_ITERATIONS {
        let frame = FrameBuilder::build_binary(&payload, true, true);
        let mut parser = Parser::new();
        if parser.parse(&frame).0 != ParseResult::Complete {
            eprintln!("masked parse failed");
            std::process::exit(1);
        }
    }
    let rt_sec = ws_rs_benches::common::elapsed_sec(roundtrip_start, Instant::now());
    print_throughput(
        "masked_build_parse",
        PAYLOAD_SIZE as f64 * ROUNDTRIP_ITERATIONS as f64,
        rt_sec,
        ROUNDTRIP_ITERATIONS,
    );
}
